Remote Connections #16

Jonatan-B opened the pull request:
## What Changed?

In order to get this working I had to made some changes on the WMI calls, as well as the File Paths returned for the service.

Added a new private property to store the name of the server we'll be connecting to.

Modified the GetService()  call to request the information to the remote machine, and added wrapped it in a try/catch to catch access/connection issues.

Added new elements to the design to get input from the user for the server connection, and methods to handle connections, and disconnection events.

Added a statusstrip to easy to see show connection information, and modified the title to show only Total: 0 if disconnected, as it would be weirdly formatted, 'Total: 0,', if it was disconnected.

## Screenshots
![image](https://user-images.githubusercontent.com/11668742/33634160-dfae04fe-d9d8-11e7-9eaf-03f5e60312a6.png)

![image](https://user-images.githubusercontent.com/11668742/33634195-fa387d7c-d9d8-11e7-8d94-4685777cebad.png)



## Closed issues
Remote Connections #16

Jonatan-B committed:
Added support to use the application with remote machines

In order to get this working I had to made some changes on the WMI calls, as well as the File Paths returned for the service.

Added a new private property to store the name of the server we'll be connecting to.

Modified the GetService()  call to request the information to the remote machine, and added wrapped it in a try/catch to catch access/connection issues.

Added new elements to the design to get input from the user for the server connection, and methods to handle connections, and disconnection events.

Added a statusstrip to easy to see show connection information, and mofied the title to show only Total: 0 if disconnected, as it would be weirdly formatted, 'Total: 0,', if it was disconnected.

BlythMeister commented on src/ServiceBouncer/MainForm.cs:
```diff
@@ -8,13 +8,17 @@
         private readonly BindingList<ServiceViewModel> services = new SortableBindingList<ServiceViewModel>();
         private bool isActive;
+        private string _machineHostname = "localhost";
+
```
There doesn't seem to be a need to have this property and field.
Since both are private, just use the field.
Could you also initialise it in the ctor rather than inline

BlythMeister commented on src/ServiceBouncer/MainForm.cs:
```diff
@@ -137,20 +141,62 @@ private void InstallClicked(object sender, EventArgs e)
-            foreach (var model in systemServices.Select(service => new ServiceViewModel(service)).OrderBy(x => x.Name))
+            try
+            {
+                var systemServices = await Task.Run(() => ServiceController.GetServices(MachineHostname).Where(service => service.DisplayName.IndexOf(toolStripFilterBox.Text, StringComparison.OrdinalIgnoreCase) >= 0));
```
How does this handle authentication if your current running user isn't a domain user, or the machine being connected to is on a different domain.

Could we add a user/password entry option for this case?

Jonatan-B replied:
Currently it's all based on the running user. I could change it to be able to provide credentials though. I'm wondering if it might just be the same to use RunAs on the program though?

BlythMeister replied:
I think that adding ability to run connect as a different user will mean someone doesn't have to keep disconnecting should they only have local users and not a domain user to connect as.

BlythMeister replied:
Logic like this will allow for impresonation: https://stackoverflow.com/a/212673
We could prompt user to ask if a different login user is required, and if so, show a login form which has the information pushed into the impersonate?

BlythMeister commented on src/ServiceBouncer/MainForm.cs:
```diff
@@ -137,20 +141,62 @@ private void InstallClicked(object sender, EventArgs e)
+            if (services.Any())
+            {
+                var titles = services.GroupBy(s => s.Status).Select(s => (string.IsNullOrWhiteSpace(s.Key) ? "Unknown" : s.Key) + ": " + s.Count());
+                Text = "Total: " + services.Count + titles.Aggregate((sentance, next) => sentance + ", " + next );
```
This has changed the formatting as there is no longer a space after the total count

BlythMeister replied:
Also, (though a micro optimisation) using aggregate here will create lots of strings in memory and then return the final string.
String.join does not do this, so can this side of the new if statement remain unchanged

BlythMeister reviewed (changes requested):
Thank you for your contribution.
There are a few requested changes.
Can you also look at the vertical alignment of the new toolstrip components.
Existing text and boxes have top padding to align center.

Finally, the filter box (on the right) seems to have vanished on your screenshots.  I'm assuming the strip is too wide at the minimum width for the buttons which we should fix.

Jonatan-B committed:
Fixed an issue with the toolstrip pushing the filter textbox and icon out of view by moving the connect option into the primary toolstrip instead of placing it on its own.

Removes the MachineHostname property and refactor the class to simply use the private field, and initiated it from the Constructor.

Refactor the SetTitle method to not use Aggregate and use STring.Join() instead.

Jonatan-B commented:
Hello BythMeister, 

I tried to make all of the requested changes. The only thing that I wasn't able to replicate was the padding on the text field.

> Existing text and boxes have top padding to align center.

I wasn't able to see it in the filter text field properties, am I missing something?

BlythMeister commented:
@Jonatan-B i will checkout your branch and tweak it for you :)

BlythMeister committed:
Function re-order and rename for consistency

BlythMeister committed:
Fix wording on connect/disconnect button

BlythMeister commented:
@Jonatan-B made a couple of changes, but i was wrong about the padding, apologies.

I'm happy to marge this, and will open a fresh issue regarding the ability to "log in" when connecting remotely.
